### ISSUE TITLE
6x Performance Drop

### ISSUE THREAD

**Issue by nwoltman** [mcollina/sonic-boom#22]
The `benchSonicSync*1000` benchmark is about 6x slower after the commit 2593841172d67c564a77660b30d8a693efa01757 ("Sync writing must be fully sync").

Before 2593841172d67c564a77660b30d8a693efa01757

```
benchSonic*1000: 3338.809ms
benchSonicSync*1000: 3489.371ms
benchSonic4k*1000: 3302.510ms
benchSonicSync4k*1000: 3489.236ms
benchCore*1000: 10764.431ms
benchSonic*1000: 3301.564ms
benchSonicSync*1000: 3490.448ms
benchSonic4k*1000: 3229.361ms
benchSonicSync4k*1000: 3509.415ms
benchCore*1000: 10422.195ms
```

After 2593841172d67c564a77660b30d8a693efa01757

```
benchSonic*1000: 3391.855ms
benchSonicSync*1000: 20495.038ms  // was 3489.371ms
benchSonic4k*1000: 4693.662ms
benchSonicSync4k*1000: 3256.744ms
benchCore*1000: 8517.529ms
benchSonic*1000: 3364.862ms
benchSonicSync*1000: 20543.356ms  // was 3490.448ms
benchSonic4k*1000: 5386.639ms
benchSonicSync4k*1000: 3167.034ms
benchCore*1000: 8749.439ms
```

Is it still ok to use `pino.destination` in production? This benchmark makes me worried since `pino.destination` creates a sonic-boom instance the same way as the benchmark that dropped in performance: `new SonicBoom(fd, 0, true)`.

Also, I think the `benchSonicSync4k*1000` benchmark may still be fast because it is not fully sync (possibly due to [this code](https://github.com/mcollina/sonic-boom/blob/f41da47b1158cdd18debe7ad2b22fc1ca193462f/index.js#L70-L74)).

**Comment by mcollina** [comment]
Which operating system are you using? Which version of Node.js? That should not cause a perf drop.

**Comment by nwoltman** [comment]
Ubuntu 18.04
Node 11.6.0

**Comment by mcollina** [comment]
The answer to this is complex, but do not worry as pino is doing the right thing with the default. Before that change, `pino.destination(1)` could have lost or corrupt log lines on broken exits. Note that using standard `process.stdout` could lead to the same corrupted log lines in case of processes within a pipe (in other terms, that is also a bug in Node.js core, I can find a reference if you want one). We made the call that `pino.destination(1)` should be *safe*.

Node that for maximum throughput `pino.extreme(1)` is definitely recommened in combination with https://getpino.io/#/docs/api?id=pino-final.

In #23, I've fixed a bad bottleneck/memory issue condition that the current implementation had, and I also fixed the issue you noted in https://github.com/mcollina/sonic-boom/blob/f41da47b1158cdd18debe7ad2b22fc1ca193462f/index.js#L70-L74. Sync mode with buffering is actually faster in this benchmark because it allocates less memory by blocking the main JS thread. However that gain is not reflected in real-life situations.
